bokehjs/src/lib/models/annotations/color_bar.ts

```diff
@@ -121,475 +249,279 @@ export class ColorBarView extends AnnotationView {
-          sy = (vr.end + vr.start)/2 - legend_height/2
-          break
-        case 'center_right':
-          sx = hr.end - legend_margin - legend_width
-          sy = (vr.end + vr.start)/2 - legend_height/2
-          break
-      }
-    } else if (isArray(location) && location.length == 2) {
-      const [vx, vy] = location
-      sx = panel.xview.compute(vx)
-      sy = panel.yview.compute(vy) - legend_height
-    } else
-      unreachable()
-
-    return {sx, sy}
-  }
-
-  protected _render(): void {
-    const {ctx} = this.layer
-    ctx.save()
-
-    const {sx, sy} = this.compute_legend_location()
-    ctx.translate(sx, sy)
-    this._draw_bbox(ctx)
-
-    const image_offset = this._get_image_offset()
-    ctx.translate(image_offset.x, image_offset.y)
-
-    this._draw_image(ctx)
-
-    const tick_info = this.tick_info()
-    this._draw_major_ticks(ctx, tick_info)
-    this._draw_minor_ticks(ctx, tick_info)
-    this._draw_major_labels(ctx, tick_info)
-
-    if (this.model.title)
-      this._draw_title(ctx)
-
-    ctx.restore()
   }
 
-  protected _draw_bbox(ctx: Context2d): void {
-    const bbox = this.compute_legend_dimensions()
-    ctx.save()
-    if (this.visuals.background_fill.doit) {
-      this.visuals.background_fill.set_value(ctx)
-      ctx.fillRect(0, 0, bbox.width, bbox.height)
-    }
-    if (this.visuals.border_line.doit) {
-      this.visuals.border_line.set_value(ctx)
-      ctx.strokeRect(0, 0, bbox.width, bbox.height)
-    }
-    ctx.restore()
-  }
-
-  protected _draw_image(ctx: Context2d): void {
-    const image = this._computed_image_dimensions()
-    ctx.save()
-    ctx.setImageSmoothingEnabled(false)
-    ctx.globalAlpha = this.model.scale_alpha
-    ctx.drawImage(this.image, 0, 0, image.width, image.height)
-    if (this.visuals.bar_line.doit) {
-      this.visuals.bar_line.set_value(ctx)
-      ctx.strokeRect(0, 0, image.width, image.height)
-    }
-    ctx.restore()
-  }
-
-  protected _draw_major_ticks(ctx: Context2d, tick_info: TickInfo): void {
-    if (!this.visuals.major_tick_line.doit)
-      return
+  update_layout(): void {
+    const {location, width: w, height: h, padding, margin} = this.model
+
+    const [valign, halign] = (() => {
+      if (isString(location)) {
+        switch (location) {
+          case "top_left":
+            return ["start", "start"] as const
+          case "top_center":
+            return ["start", "center"] as const
+          case "top_right":
+            return ["start", "end"] as const
+          case "bottom_left":
+            return ["end", "start"] as const
+          case "bottom_center":
+            return ["end", "center"] as const
+          case "bottom_right":
+            return ["end", "end"] as const
+          case "center_left":
+            return ["center", "start"] as const
+          case "center":
+            return ["center", "center"] as const
+          case "center_right":
+            return ["center", "end"] as const
```

Comment:
couldn't all of this huge switch be
```javascript
return location.replace(/bottom|right/, "end").replace(/top|left/, "start")
```

Reply:
No, because this maps types as well and `"center"` would need a special case. Besides, this code will be moved out of color bar when this part of layout is generalized, so when legend is redesigned.